hikari/polls.py

```diff
@@ -0,0 +1,325 @@
+    """Whether the current user voted for this answer."""
+
+
+class PollLayoutType(int, enums.Enum):
+    """Layout of a poll."""
+
+    DEFAULT = 1
+    """The default layout of a poll."""
+
+
+class PartialPoll:
+    """Base class for all poll objects."""
+
+    __slots__: typing.Sequence[str] = ("_question", "_answers", "_allow_multiselect", "_layout_type", "_counter")
+
+    def __init__(self, question: str, allow_multiselect: bool, layout_type: typing.Union[int, PollLayoutType]):
+        self._question = PollMedia(text=question)  # Only text is supported for question
+        self._allow_multiselect = allow_multiselect
+        self._layout_type = layout_type
+
+        # Answer is required, but we want users to user add_answer() instead of
+        # providing at initialization.
+        #
+        # Considering that answer ID can be arbitrary, `list`-based approaches
+        # like that of hikari.embeds.Embed._fields, while feasible to implement,
+        # would decrease long-term maintainability. I'm opting to use a `dict`
+        # here to simplify the implementation with some performance trade-off
+        # due to hashmap overhead.
+        self._answers: typing.MutableMapping[int, PollAnswer] = {}  # TODO: Do we need to set to None?
+
+        # Counter to keep track of the answer IDs
+        #
+        # Discord (at the time of writing this) starts at 1, I'm opting to follow
+        # the behaviour.
+        self._counter = 1
+
+    @property
+    def question(self) -> PollMedia:
+        """Returns the question of the poll."""
+        return self._question
+
+    @question.setter
+    def question(self, value: str) -> None:
+        self._question = PollMedia(text=value)
+
+    @property
+    def allow_multiselect(self) -> bool:
+        """Returns whether the poll allows multiple answers."""
+        return self._allow_multiselect
+
+    @allow_multiselect.setter
+    def allow_multiselect(self, value: bool) -> None:
+        self._allow_multiselect = value
+
+    @property
+    def layout_type(self) -> PollLayoutType:
+        """Returns the layout type of the poll."""
+        return PollLayoutType(self._layout_type)
+
+    @layout_type.setter
+    def layout_type(self, value: typing.Union[int, PollLayoutType]) -> None:
+        self._layout_type = value
+
+    @property
+    def answers(self) -> typing.MutableMapping[int, PollAnswer]:
+        """Returns the answers of the poll.
+
+        !!! note
+            Use [`hikari.polls.Poll.add_answer`][] to add a new answer,
+            [`hikari.polls.Poll.edit_answer`][] to edit an existing answer, or
+            [`hikari.polls.Poll.remove_answer`][] to remove a answer.
+        """
+        return self._answers
+
+
+class PollCreate(PartialPoll):
+    """Used to create a poll."""  # TODO: Improve this docstring
+
+    __slots__: typing.Sequence[str] = ("_duration",)
+
+    def __init__(
+        self,
+        question: str,
+        duration: int,
```

Comment:
It would be nice to be able to set the duration via a `datetime.timedelta` object at minimum, and maybe support a `datetime.datetime` object, that strips of `datetime.datetime.now()`? not sure about the second one, but definitely the first.

Reply:
A `datetime.timedelta` does sound better than just an `int`, but the problem is Discord's API only allows for `int` up to 144hrs (7 days).
![image](https://github.com/hikari-py/hikari/assets/31789326/2a6fe08f-5a34-4816-8699-ba14f1ba0d0c)
We could just do `floor(datetime.timedelta.hours)` but that feels like it's just `int` with extra steps.